stackslib/src/chainstate/nakamoto/coordinator/mod.rs

```diff
@@ -227,125 +217,99 @@ pub fn get_nakamoto_reward_cycle_info<U: RewardSetProvider>(
-            "anchor_chain_tip_height" => %parent_block_header.burn_header_height,
-            "first_prepare_sortition_id" => %first_sortition_id
-        );
+    let anchor_block_sn = SortitionDB::get_block_snapshot_consensus(
+        sort_db.conn(),
+        &anchor_block_header.consensus_hash,
+    )?
+    .expect("FATAL: no snapshot for winning PoX anchor block");
+
+    // make sure the `anchor_block` field is the same as whatever goes into the block-commit,
+    // or PoX ancestry queries won't work
+    let (block_id, stacks_block_hash) = match anchor_block_header.anchored_header {
+        StacksBlockHeaderTypes::Epoch2(ref header) => (
+            StacksBlockId::new(&anchor_block_header.consensus_hash, &header.block_hash()),
+            header.block_hash(),
+        ),
+        StacksBlockHeaderTypes::Nakamoto(ref header) => {
+            (header.block_id(), BlockHeaderHash(header.block_id().0))
+        }
+    };
 
-        let reward_set = provider.get_reward_set(
-            reward_start_height,
-            chain_state,
-            burnchain,
-            sort_db,
-            &block_id,
-        )?;
-        debug!(
-            "Stacks anchor block (ch {}) {} cycle {} is processed",
-            &anchor_block_header.consensus_hash, &block_id, reward_cycle
-        );
-        let anchor_status =
-            PoxAnchorBlockStatus::SelectedAndKnown(stacks_block_hash, txid, reward_set);
+    let txid = anchor_block_sn.winning_block_txid;
+
+    info!(
+        "Anchor block selected";
+        "cycle" => reward_cycle,
+        "block_id" => %block_id,
+        "consensus_hash" => %anchor_block_header.consensus_hash,
+        "burn_height" => anchor_block_header.burn_header_height,
+        "anchor_chain_tip" => %anchor_block_header.index_block_hash(),
+        "anchor_chain_tip_height" => %anchor_block_header.burn_header_height,
+        "first_prepare_sortition_id" => %first_sortition_id
+    );
 
-        let rc_info = RewardCycleInfo {
-            reward_cycle,
-            anchor_status,
-        };
+    let reward_set = provider.get_reward_set(
+        reward_start_height,
+        chain_state,
+        burnchain,
+        sort_db,
+        &block_id,
+    )?;
+    debug!(
+        "Stacks anchor block (ch {}) {} cycle {} is processed",
+        &anchor_block_header.consensus_hash, &block_id, reward_cycle
+    );
+    let anchor_status = PoxAnchorBlockStatus::SelectedAndKnown(stacks_block_hash, txid, reward_set);
 
-        // persist this
-        let mut tx = sort_db.tx_begin()?;
-        SortitionDB::store_preprocessed_reward_set(&mut tx, &first_sortition_id, &rc_info)?;
-        tx.commit()?;
+    let rc_info = RewardCycleInfo {
+        reward_cycle,
+        anchor_status,
+    };
 
-        return Ok(Some(rc_info));
-    }
+    // persist this
+    let mut tx = sort_db.tx_begin()?;
+    SortitionDB::store_preprocessed_reward_set(&mut tx, &first_sortition_id, &rc_info)?;
```

Comment:
Don't we want to persist this first and foremost in the Stacks chainstate?  If so, then would it make more sense to just load the persisted reward set out of the chainstate, instead of recalculating it?

Reply:
Yeah -- that's more or less the `TODO` item I've left open. My current understanding of how things should work:

1. `.signers` update occurs in the setup step of the first block in a prepare phase. This update writes the signers stackerdb data (currently implemented), _and_ the reward set state to `.signers` (not implemented).
2. `get_reward_set_nakamoto()` reads the reward set state directly from `.signers`, performing no calculations: its just essentially loading a vec.

This code you commented on would stay the same: the "reward set provider" is still invoked, just what it is doing is different. And the first time its invoked, it stashes its results in the SortitionDB as well so that it can be read directly from there.

The way it works now is that (1) and (2) independently calculate the reward set. That's not tenable because it introduces the possibility of different results even without a bug: one calculates at the start of the block and the other calculates at the end of the same block. But, I do want to leave it as a TODO, because I can handle it in a follow-up PR while unblocking the signer/miner communication work.

Reply:
Ah I see you have that commented above.  Can you please open a ticket for it?

Reply:
Lol we both commented at the same time.

Agreed with the above.  Yeah, a follow-on PR is fine.